### ISSUE TITLE
Bug Report - Sometimes jrnl crashes and truncates journal file

### ISSUE THREAD

**Issue by kjanezic95** [jrnl-org/jrnl#1599]
### Diagnostic output

jrnl: v2.8.3
Python: 3.8.10 (default, Jun 22 2022, 20:18:18)
[GCC 9.4.0]
OS: Linux 5.10.16.3-microsoft-standard-WSL2

### Current Behavior

Sometimes (2 times this month), jrnl crashes and just truncates whole journal file. 

This is the actual error:  ("JRNL $ " is just a prompt that sends input to "jrnl $prompt")
```sh
JRNL $ @tag Something. Something else: and also something
[Entry added to default journal]
Traceback (most recent call last):
  File "/usr/local/bin/jrnl", line 8, in <module>
    sys.exit(cli())
  File "/usr/local/lib/python3.8/dist-packages/jrnl/cli.py", line 34, in cli
    return run(args)
  File "/usr/local/lib/python3.8/dist-packages/jrnl/jrnl.py", line 71, in run
    write_mode(**kwargs)
  File "/usr/local/lib/python3.8/dist-packages/jrnl/jrnl.py", line 150, in write_mode
    journal.write()
  File "/usr/local/lib/python3.8/dist-packages/jrnl/Journal.py", line 99, in write
    self._store(filename, text)
  File "/usr/local/lib/python3.8/dist-packages/jrnl/Journal.py", line 341, in _store
    f.write(text)
UnicodeEncodeError: 'utf-8' codec can't encode character '\udcc5' in position 28563: surrogates not allowed
```

### Expected Behavior

Journal files should not get truncated, even if jrnl crashes.

### Repro Steps

I cannot reproduce it. Trying the same entry does not crash jrnl.

### Debug output

<details><summary>Debug output</summary>

```
YYYYYY@XXXXXX:[~]$ jrnl --debug @test test. test: test
DEBUG    root         Parsed args: Namespace(config_override=[], contains=None, day=None, debug=True, delete=False, edit=False, end_date=None, excluded=[], export=False, filename=None, limit=None, month=None, on_date=None, postconfig_cmd=None, preconfig_cmd=None, short=False, starred=False, start_date=None, strict=False, tags=False, text=['@test', 'test.', 'test:', 'test'], today_in_history=False, year=None)
DEBUG    root         Reading configuration from file /home/YYYYYYY/.config/jrnl/jrnl.yaml
DEBUG    root         Using configuration "{'colors': {'body': 'none', 'date': 'YELLOW', 'tags': 'RED', 'title': 'none'}, 'default_hour': 9, 'default_minute': 0, 'editor': 'vim', 'encrypt': False, 'highlight': True, 'indent_character': '|', 'journals': {'default': '/mnt/KristjanShare/jrnl/jrnl.txt'}, 'linewrap': 79, 'tagsymbols': '@', 'template': False, 'timeformat': '%Y-%m-%d %H:%M', 'version': 'v2.8.3'}"
DEBUG    root         Using journal name: default
DEBUG    root         opened PlainJournal with 461 entries
DEBUG    root         Write mode: starting
DEBUG    root         Write mode: cli text detected: ['@test', 'test.', 'test:', 'test']
DEBUG    root         Write mode: appending raw text to journal "default": @test test. test: test
[Entry added to default journal]
--- Logging error ---
Traceback (most recent call last):
  File "/usr/lib/python3.8/logging/__init__.py", line 1085, in emit
    msg = self.format(record)
  File "/usr/lib/python3.8/logging/__init__.py", line 929, in format
    return fmt.format(record)
  File "/usr/lib/python3.8/logging/__init__.py", line 668, in format
    record.message = record.getMessage()
  File "/usr/lib/python3.8/logging/__init__.py", line 373, in getMessage
    msg = msg % self.args
TypeError: not all arguments converted during string formatting
Call stack:
  File "/usr/local/bin/jrnl", line 8, in <module>
    sys.exit(cli())
  File "/usr/local/lib/python3.8/dist-packages/jrnl/cli.py", line 34, in cli
    return run(args)
  File "/usr/local/lib/python3.8/dist-packages/jrnl/jrnl.py", line 71, in run
    write_mode(**kwargs)
  File "/usr/local/lib/python3.8/dist-packages/jrnl/jrnl.py", line 151, in write_mode
    logging.debug("Write mode: completed journal.write()", args.journal_name, raw)
Message: 'Write mode: completed journal.write()'
Arguments: ('default', '@test test. test: test')

### Other Information

"JRNL $ " is just a prompt that sends input to "jrnl $prompt"

Im addidng this example, although I do not think this is the issue:

 case "$prompt" in
      h|help) help ;;
      d|delete) delete ;;
      e|edit) edit ;;
      r|refresh) jrnl -n25 ;;
      s|search) search ;;
      t|tag) tag ;;
      l|list) list ;;
      exit) exit 0 ;;
      *) jrnl "$prompt" && jrnl -n25 ;;
  esac

```

</details>

**Comment by micahellison** [comment]
Hi @kjanezic95, thanks so much for filing this. We're looking into it and are still not sure what the root cause is, though here's what we can tell so far: jrnl only supports UTF-8, and there was something in either your journal or your call to jrnl that somehow uses UTF-16 characters (specifically, [surrogates](https://learn.microsoft.com/en-us/globalization/encoding/surrogate-pairs)).

I don't know how this got there -- maybe jrnl can write them but just can't read them, or maybe you edited your journal file manually to add them in. Maybe they somehow got piped to jrnl from your script? I wonder if perhaps you entered one of these characters in your special jrnl prompt, then pressed backspace and it didn't fully remove the character. Or maybe some keyboard shortcut that inserts these characters.

Whatever the case, I did noticed that your version numbers are a bit out of date. Could you try updating Python to 3.9 or 3.10, then updating jrnl to the latest version and running again? That might help.

**Comment by kjanezic95** [comment]
Hello, from what you say it is probably the scripts fault, as jrnl never did that when I used it "normally". I am suspecting I somehow inserted one of this characters when I was editing the prompt - deleting text, moving around, etc...

I am using jrnl in WSL Ubuntu 20 and Python is updated to the most recent version. I did try to upgrade manually but then I had 2 Pythons. When I removed old one everything Python related just collapsed (pip, pipx, jrnl, etc...). 

For now I will stay with versions that are available by default in the WSL Ubuntu 20 and backup the journal file more often :) 

Thank you for your help. As I will not upgrade Python, this case can get closed.
